src/python/pants/engine/process.py

```diff
@@ -482,15 +485,26 @@ def __init__(
 
 @rule(desc="Find binary path", level=LogLevel.DEBUG)
 async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
-    # TODO(John Sirois): Replace this script with a statically linked native binary so we don't
+    # If we are not already locating bash, recurse to locate bash to use it as an absolute path in
+    # our shebang. This avoids mixing locations that we would search for bash into the search paths
+    # of the request we are servicing.
+    # TODO(#10769): Replace this script with a statically linked native binary so we don't
     #  depend on either /bin/bash being available on the Process host.
+    if request.binary_name == "bash":
+        shebang = "#!/usr/bin/env bash"
+    else:
+        bash_request = BinaryPathRequest(binary_name="bash", search_path=BASH_SEARCH_PATHS)
+        bash_paths = await Get(BinaryPaths, BinaryPathRequest, bash_request)
+        if not bash_paths.first_path:
+            raise BinaryNotFoundError(bash_request, rationale="use it to locate other executables")
+        shebang = f"#!{bash_paths.first_path.path}"
 
     # Note: the backslash after the """ marker ensures that the shebang is at the start of the
     # script file. Many OSs will not see the shebang if there is intervening whitespace.
     script_path = "./script.sh"
```

Comment:
Nit: could you please update this to be `find_binary.sh`? I think it will be more informative for end users in error messages.

(Ik it was this way before)